proxy/src/main/java/com/wavefront/agent/listeners/ZipkinPortUnificationHandler.java

```diff
@@ -0,0 +1,262 @@
+      logger.log(Level.WARNING, "Zipkin batch processing failed", Throwables.getRootCause(e));
+    }
+    writeHttpResponse(ctx, status, output, incomingRequest);
+  }
+
+  private void processZipkinSpans(List<zipkin2.Span> zipkinSpans) {
+    for (zipkin2.Span zipkinSpan : zipkinSpans) {
+      processZipkinSpan(zipkinSpan);
+    }
+  }
+
+  private void processZipkinSpan(zipkin2.Span zipkinSpan) {
+    List<Annotation> annotations = addAnnotations(zipkinSpan);
+
+    /** Add source of the span following the below:
+     *    1. If "source" is provided by span tags , use it else
+     *    2. Set "source" to local service endpoint's ipv4 address, else
+     *    3. Default "source" to "zipkin".
+     */
+    String sourceName = DEFAULT_SOURCE;
+    if (zipkinSpan.tags().get(SOURCE_KEY) != null) {
+      sourceName = zipkinSpan.tags().get(SOURCE_KEY);
+    } else if (zipkinSpan.localEndpoint().ipv4() != null) {
+      sourceName = zipkinSpan.localEndpoint().ipv4();
+    }
+
+    //Build wavefront span
+    Span newSpan = Span.newBuilder().
+        setCustomer("dummy").
+        setName(zipkinSpan.name()).
+        setSource(sourceName).
+        setSpanId(getSpanUuid(zipkinSpan)).
+        setTraceId(Utils.convertToUuidString(zipkinSpan.traceId())).
+        setStartMillis(zipkinSpan.timestampAsLong() / 1000).
+        setDuration(zipkinSpan.durationAsLong() / 1000).
+        setAnnotations(annotations).
+        build();
+
+    // Log Zipkin spans as well as Wavefront spans for debugging purposes.
+    if (zipkinDataLogger.isLoggable(Level.FINEST)) {
+      zipkinDataLogger.info("Inbound Zipkin span: " + zipkinSpan.toString());
+      zipkinDataLogger.info("Converted Wavefront span: " + newSpan.toString());
+    }
+
+    handler.report(newSpan);
+  }
+
+  private List<Annotation> addAnnotations(zipkin2.Span zipkinSpan) {
+    List<Annotation> annotations = new ArrayList<>();
+
+    // Set Span's Application Tags.
+    String applicationName = zipkinSpan.tags().get(APPLICATION_KEY) == null ? DEFAULT_APPLICATION
+        : zipkinSpan.tags().get(APPLICATION_KEY);
+    annotations.add(new Annotation(APPLICATION_KEY, applicationName));
+
+    annotations.add(new Annotation(SERVICE_KEY, zipkinSpan.localServiceName()));
+
+    if (zipkinSpan.tags().get(CLUSTER_KEY) != null) {
+      annotations.add(new Annotation(CLUSTER_KEY, zipkinSpan.tags().get(CLUSTER_KEY)));
+    }
+    if (zipkinSpan.tags().get(SHARD_KEY) != null) {
+      annotations.add(new Annotation(SHARD_KEY, zipkinSpan.tags().get(SHARD_KEY)));
+    }
+
+    // Set Span's References.
+    if (zipkinSpan.parentId() != null) {
+      annotations.add(new Annotation(TraceConstants.PARENT_KEY, Utils.convertToUuidString(zipkinSpan
+          .parentId())));
+    }
+
+    // Set Span Http Tags.
+    if (zipkinSpan.tags() != null && zipkinSpan.tags().size() > 0) {
+      annotations.add(new Annotation("span.kind", zipkinSpan.kind().toString().toLowerCase()));
```

Comment:
It appears `kind()` can return null: https://github.com/openzipkin/zipkin/blob/0efe65a8685c01d2b09da20c94667fb2c07d21c4/zipkin/src/main/java/zipkin2/Span.java#L131

Please handle accordingly to prevent any NPEs.